grid-proxy/internal/gpuindexer/gpuindexer.go

```diff
@@ -0,0 +1,176 @@
+	checkInterval   time.Duration
+	workerBatchSize int
+}
+
+func NewNodeGPUIndexer(
+	ctx context.Context,
+	relayURL,
+	mnemonics string,
+	sub *substrate.Substrate,
+	db db.Database,
+	indexerCheckIntervalMins int,
+	workerBatchSize int) (*NodeGPUIndexer, error) {
+	indexer := &NodeGPUIndexer{
+		db:              db,
+		nodesGPUChan:    make(chan []types.NodeGPU),
+		checkInterval:   time.Duration(indexerCheckIntervalMins) * time.Minute,
+		workerBatchSize: workerBatchSize,
+	}
+
+	client, err := direct.NewClient(ctx, direct.KeyTypeSr25519, mnemonics, relayURL, "tfgrid_proxy_indexer", sub, true, indexer.relayCallback)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create direct RMB client: %w", err)
+	}
+	indexer.relayClient = client
+
+	return indexer, nil
+}
+
+func (n *NodeGPUIndexer) worker(ctx context.Context) {
+	var nodesGPUBuffer []types.NodeGPU
+	ticker := time.NewTicker(workerCleanupInterval)
+	for {
+		select {
+		case nodesGPU := <-n.nodesGPUChan:
+			nodesGPUBuffer = append(nodesGPUBuffer, nodesGPU...)
+			if len(nodesGPUBuffer) >= n.workerBatchSize {
+				log.Debug().Msg("flushing gpu indexer buffer")
+				err := n.db.UpsertNodesGPU(nodesGPUBuffer)
+				if err != nil {
+					log.Error().Err(err).Msg("failed to update GPU info in GPU indexer")
+					continue
+				}
+				nodesGPUBuffer = nil
+			}
+		// This case should only be triggered when there leftovers data in the buffer
+		case <-ticker.C:
+			if len(nodesGPUBuffer) != 0 {
+				log.Debug().Msg("cleaning up gpu indexer buffer")
+				err := n.db.UpsertNodesGPU(nodesGPUBuffer)
+				if err != nil {
+					log.Error().Err(err).Msg("failed to update GPU info in GPU indexer")
+					continue
+				}
+				nodesGPUBuffer = nil
+			}
+		case <-ctx.Done():
+			log.Error().Err(ctx.Err()).Msg("worker exited")
+			return
+
+		}
+	}
+}
+
+func (n *NodeGPUIndexer) relayCallback(response *rmbTypes.Envelope, callBackErr error) {
+
+	errResp := response.GetError()
+
+	if errResp != nil {
+		log.Error().Msg(errResp.Message)
+		return
+	}
+
+	resp := response.GetResponse()
+	if resp == nil {
+		log.Error().Msg("received a non response envelope")
+		return
+	}
+
+	if response.Schema == nil || *response.Schema != rmb.DefaultSchema {
+		log.Error().Msgf("invalid schema received expected '%s'", rmb.DefaultSchema)
+		return
+	}
+
+	output := response.Payload.(*rmbTypes.Envelope_Plain).Plain
+
+	var nodesGPU []types.NodeGPU
+	err := json.Unmarshal(output, &nodesGPU)
+	if err != nil {
+		log.Error().Err(err).Msg("failed to unmarshal GPU information response")
+		return
+
+	}
+	for i := range nodesGPU {
+		nodesGPU[i].NodeTwinID = response.Source.Twin
+	}
+	// Will be using only one worker giving the current amount of nodes supporting GPUs
+	if len(nodesGPU) != 0 {
+		n.nodesGPUChan <- nodesGPU
+	}
+
+}
+
+func (n *NodeGPUIndexer) Start(ctx context.Context) {
+	ticker := time.NewTicker(n.checkInterval)
+	log.Info().Msg("GPU indexer started")
+	go n.worker(ctx)
+	n.run(ctx)
```

Comment:
looks good, but isn't it more clean if we did it like 
```
go n.worker(ctx)
go n.run(ctx) and inside run we do the ticker?

Reply:
This is done this way so that `run` runs at the start of the code, otherwise it will have to wait for the interval to pass before starting the indexer job.

Reply:
> This is done this way so that `run` runs at the start of the code, otherwise it will have to wait for the interval to pass before starting the indexer job.

yeah. np for that I mean this to be done inside the run method